File: src/core/scala/mrtjp/projectred/compatibility/ComputerCraftCompatibility.scala
package mrtjp.projectred.compatibility

import dan200.computercraft.api.ComputerCraftAPI
import dan200.computercraft.api.redstone.IBundledRedstoneProvider
import mrtjp.projectred.api.{IBundledTileInteraction, ProjectRedAPI}
import mrtjp.projectred.core.BundledCommons
import net.minecraft.util.Direction
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

/**
 * Provides bundled cable compatibility with ComputerCraft
 */
object ComputerCraftCompatibility
{
    def init():Unit = {
        if (ProjectRedAPI.transmissionAPI != null) { // Check if Transmission is installed
            println("Loading ProjectRed compatibility: ComputerCraft")
            ComputerCraftAPI.registerBundledRedstoneProvider(new CCPRBundledRedstoneProvider)
            ProjectRedAPI.transmissionAPI.registerBundledTileInteraction(new PRCCBundledTileInteraction)
        }
    }

    /**
     * This is used by ComputerCraft to query bundled signals from third-party entities
     */
    class CCPRBundledRedstoneProvider extends IBundledRedstoneProvider
    {
        override def getBundledRedstoneOutput(world:World, pos:BlockPos, side:Direction):Int = {
            val sig = ProjectRedAPI.transmissionAPI.getBundledInput(world, pos.relative(side), side.getOpposite)
            BundledCommons.packDigital(sig)
        }
    }

    /**
     * On the other side, this is used by ProjectRed to query bundled signals from third-party entities
     */
    class PRCCBundledTileInteraction extends IBundledTileInteraction
    {
        override def isValidInteractionFor(world:World, pos:BlockPos, side:Direction):Boolean =
            ComputerCraftAPI.getBundledRedstoneOutput(world, pos, side) > -1

        override def canConnectBundled(world:World, pos:BlockPos, side:Direction):Boolean = true

        override def getBundledSignal(world:World, pos:BlockPos, side:Direction):Array[Byte] = {
            val sig = ComputerCraftAPI.getBundledRedstoneOutput(world, pos, side)
            BundledCommons.unpackDigital(null, sig)
        }
    }
}

File: core/src/main/java/mrtjp/projectred/compatibility/ComputerCraftCompatibility.java
package mrtjp.projectred.compatibility;

import dan200.computercraft.api.ComputerCraftAPI;
import dan200.computercraft.api.redstone.BundledRedstoneProvider;
import mrtjp.projectred.api.IBundledTileInteraction;
import mrtjp.projectred.api.ProjectRedAPI;
import mrtjp.projectred.core.BundledSignalsLib;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.Level;

import javax.annotation.Nonnull;

import static mrtjp.projectred.core.ProjectRedCore.LOGGER;

public class ComputerCraftCompatibility {

    public static void init(Object ccModObject) {

        if (ProjectRedAPI.transmissionAPI != null) {
            LOGGER.info("Loading Project Red ComputerCraft Compatibility Module");
            ComputerCraftAPI.registerBundledRedstoneProvider(new CCPRBundledRedstoneProvider());
            ProjectRedAPI.transmissionAPI.registerBundledTileInteraction(new PRCCBundledTileInteraction());
        }
    }

    /**
     * This is used by ComputerCraft to query bundled signals from third-party entities
     */
    private static class CCPRBundledRedstoneProvider implements BundledRedstoneProvider {
        @Override
        public int getBundledRedstoneOutput(@Nonnull Level world, @Nonnull BlockPos pos, @Nonnull Direction side) {
            assert ProjectRedAPI.transmissionAPI != null;
            byte[] signal = ProjectRedAPI.transmissionAPI.getBundledInput(world, pos.relative(side), side.getOpposite());
            return BundledSignalsLib.packDigital(signal);
        }
    }

    /**
     * Used by Project Red to provide bundled connectivity to third-party blocks
     */
    private static class PRCCBundledTileInteraction implements IBundledTileInteraction {
        @Override
        public boolean isValidInteractionFor(Level world, BlockPos pos, Direction side) {
            return ComputerCraftAPI.getBundledRedstoneOutput(world, pos, side) != -1;
        }

        @Override
        public boolean canConnectBundled(Level world, BlockPos pos, Direction side) {
            return true;
        }

        @Override
        public byte[] getBundledSignal(Level world, BlockPos pos, Direction side) {
            int signal = ComputerCraftAPI.getBundledRedstoneOutput(world, pos, side);
            return BundledSignalsLib.unpackDigital(null, signal);
        }
    }
}
